### ISSUE TITLE
FastR slow to start script

### ISSUE THREAD

**Issue by peblpebl** [oracle/fastr#65]
I am running  in
VirtualBox Version 6.0.4 r128413 (Qt5.6.2)
Ubuntu 18.10 Cosmic (64bit)
a simple helloworld in fastR which takes a long time. Based on other R programs, is seems it is startup that takes the time:

<pre>
anne@ubuntu:/shared$ cat helloworld.R 
print("Hello World!")
anne@ubuntu:/shared$ time /usr/lib/graalvm/bin/Rscript --cpusampler helloworld.R 
[1] "Hello World!"
------------------------------------------------------------------------------------------------------
Sampling Histogram. Recorded 40 samples with period 1ms.
  Self Time: Time spent on the top of the stack.
  Total Time: Time spent somewhere on the stack.
  Opt %: Percent of time spent in compiled and therefore non-interpreted code.
------------------------------------------------------------------------------------------------------
 Thread: Thread[main,5,main]
 Name           |      Total Time     |  Opt % ||       Self Time     |  Opt % | Location             
------------------------------------------------------------------------------------------------------
 print.default  |         34ms  85.0% |   0.0% ||         34ms  85.0% |   0.0% | print.default.r~1-4:126-388 
 print          |         38ms  95.0% |   0.0% ||          4ms  10.0% |   0.0% | print.r~2:18-35 
 <repl wrapper> |         40ms 100.0% |   0.0% ||          2ms   5.0% |   0.0% | helloworld.R~1:0-20 
------------------------------------------------------------------------------------------------------

real	1m3.502s
user	0m58.421s
sys	0m3.515s

anne@ubuntu:/shared$ time /usr/lib/graalvm/bin/Rscript --version:graalvm
GraalVM Polyglot Engine Version 1.0.0-rc14
GraalVM Home /usr/lib/graalvm
  Installed Languages:
    R          version 3.5.1
    JavaScript version 1.0
    LLVM       version 6.0.0
    Python     version 3.7.0
  Installed Tools:
    LLVMTracerInstrument    version 1.0.0-rc14
    CPU Sampler             version 0.4.0
    CPU Tracer              version 0.3.0
    Heap Allocation Monitor version 0.1.0
    Chrome Inspector        version 0.1
    Memory Tracer           version 0.2

real	0m0.948s
user	0m1.110s
sys	0m0.147s
</pre>

It does not seem related to the virtual machine as:

<pre>
anne@ubuntu:/shared$ which Rscript
/usr/bin/Rscript
anne@ubuntu:/shared$ time Rscript helloworld.R 
[1] "Hello World!"

real	0m0.501s
user	0m0.376s
sys	0m0.096s
anne@ubuntu:/shared$ Rscript --version
R scripting front-end version 3.5.1 (2018-07-02)
</pre>

Nor that the graal is not in the PATH

<pre>
anne@ubuntu:/shared$ export PATH=/usr/lib/graalvm/bin:$PATH
anne@ubuntu:/shared$ which Rscript
/usr/lib/graalvm/bin/Rscript
anne@ubuntu:/shared$ time Rscript helloworld.R 
[1] "Hello World!"

real	0m53.571s
user	0m50.905s
sys	0m3.055s
</pre>

**Comment by steve-s** [comment]
Hi peblpebl, 

thank you for reporting your findings. 

In general, we are aware that the start-up time of FastR is worse than GNU-R and we plan to address that with an [ahead-of-time compilation of FastR](http://www.graalvm.org/docs/reference-manual/aot-compilation/) itself with the `native-image` tool. You can already try this by running `./jre/languages/R/bin/install_r_native_image`. However, some things are not yet working in the ahead-of-time compiled FastR, most notably some R packages linking with `Rcpp`. Moreover, we do not fully use the whole potential of the `native-image` tool yet.

From the user vs real time difference, it seems that you've allocated only one CPU core for your virtual machine? FastR interpreter (being Java program) is compiled on background threads by JVM and then the R code itself is compiled on background threads by Truffle Partial Evaluator (turns the interpreter into optimized machine code). That means that if you have only one HW thread, the warm-up of FastR will be much slower.

On my machine I am getting:
```
time graalvm-ce-1.0.0-rc14-dev/bin/Rscript /tmp/a.R 
[1] "Hello world"

real	0m12.011s
user	0m39.693s
sys	0m0.674s
```
vs
```
time /usr/bin/Rscript /tmp/a.R 
[1] "Hello world"

real	0m0.218s
user	0m0.137s
sys	0m0.043s
```

**Comment by peblpebl** [comment]
> In general, we are aware that the start-up time of FastR is worse than GNU-R 

I assummed as much, but did not expect 1min vs 1 sec difference, which makes me suspect that something else is going on?

> From the user vs real time difference, it seems that you've allocated only one CPU core for your virtual machine? 

I tried to give 2 cores as 1 is typically to few. VirtualBox log says
<pre>
00:00:13.833525                                Name: GenuineIntel
00:00:13.833527                            Supports: 0x00000000-0x0000000b
00:00:13.833557                              Family:  6 	Extended: 0 	Effective: 6
00:00:13.833560                               Model: 10 	Extended: 1 	Effective: 26
00:00:13.833562                            Stepping: 5
00:00:13.833563                                Type: 0 (primary)
00:00:13.833565                             APIC ID: 0x00
00:00:13.833567                        Logical CPUs: 2
</pre>

> On my machine I am getting:

Which is much more in line of what I expected.

I will experience with native image, but it is not the direction I want to go.

**Comment by peblpebl** [comment]
It seems that the start-up time that long is expected.

**Comment by steve-s** [comment]
> I assummed as much, but did not expect 1min vs 1 sec difference, which makes me suspect that something else is going on?

yes, this is indeed much more than it should be. You can try profiling it with VisualVM (included in the GraalVM distribution) and check what's taking that much time (it could be e.g. GC or compiler threads).

**Comment by steve-s** [comment]
If you could share the VirtualBox image, we can also try reproducing and debugging this on our side.

**Comment by peblpebl** [comment]
I tried to use visualvm, but could not connect while fastR was "frozen".
So I tried to see what happen on disc instead with strace -f. 
I learned 4 things:
1) graal.jar was always loaded with fd 15, and newer closed (expected).
2) There is no practically difference between (zip-) readfully and readfullyat, even though the latter explicit call seek. So all read are suffixed with a seek, even if not needed. (unexpected)
3) There is no buffer when reading the jar, so I often saw a read and the new read +100 bytes further ahead. (expected given no buffer)
4) After certain point I see a lot read calls that reads 1 byte. With a lot I mean 65000+. (unexpected).

At this point I stopped the vm, I could not repeat the 1m freeze again, more like 20-25sec.
It made no difference to place the graal.jar om ramdisk and symlinking.

So no point actually going further as I no longer can reproduce.
